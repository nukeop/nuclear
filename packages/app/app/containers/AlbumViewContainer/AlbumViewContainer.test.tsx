import { waitFor } from '@testing-library/react';
import { mountedComponentFactory, setupI18Next } from '../../../test/testUtils';
import { buildStoreState } from '../../../test/storeBuilders';

describe('Album view container', () => {
  beforeAll(() => {
    setupI18Next();
  });

  beforeEach(() => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { store } = require('@nuclear/core');
    store.clear();
  });

  it('should display an album', () => {
    const { component } = mountComponent();
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should show that an album is still loading', () => {
    const { component } = mountComponent(
      buildStoreState()
        .withPlugins()
        .withAlbumDetails({ ['test-album-id']: { loading: true } })
        .build()
    );

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should search for artist after clicking his name', async () => {
    const { component, history } = mountComponent();
    expect(history.location.pathname).toBe('/album/test-album-id');
    await waitFor(() => component.getByText(/test artist/i).click());
    expect(history.location.pathname).toBe('/artist/test-artist-id');
  });

  it('should add a single track to queue after clicking the button in the popup', async () => {
    const { component, store } = mountComponent();

    await waitFor(() => component.getByText(/test track 1/i).click());
    await waitFor(() => component.getByText(/add to queue/i).click());

    const state = store.getState();
    expect(state.queue.queueItems).toEqual([
      expect.objectContaining({
        artist: 'test artist',
        name: 'test track 1'
      })
    ]);
  });

  it('should start playing a single track after clicking the button in the popup', async () => {
    const { component, store } = mountComponent();

    await waitFor(() => component.getByText(/test track 1/i).click());
    await waitFor(() => component.getByText(/play now/i).click());

    const state = store.getState();
    expect(state.queue.queueItems).toEqual([
      expect.objectContaining({
        artist: 'test artist',
        name: 'test track 1'
      })
    ]);
    expect(state.player.playbackStatus).toEqual('PLAYING');
  });

  it('should add a single track to downloads after clicking the button in the popup', async () => {
    const { component, store } = mountComponent();

    await waitFor(() => component.getByText(/test track 1/i).click());
    await waitFor(() => component.getByText(/download/i).click());

    const state = store.getState();
    expect(state.downloads).toEqual([
      {
        completion: 0,
        status: 'Waiting',
        track: expect.objectContaining({
          artist: 'test artist',
          name: 'test track 1'
        })
      }
    ]);
  });

  it('should add album to favorites after clicking the star', async () => {
    const { component, store } = mountComponent();
    let state = store.getState();
    expect(state.favorites.albums).toEqual([]);

    await waitFor(() => component.getByTestId(/add-remove-favorite/i).click());
    state = store.getState();
    expect(state.favorites.albums).toEqual([
      expect.objectContaining({
        artist: 'test artist',
        title: 'test album'
      })
    ]);
  });

  it('should remove album from favorites after clicking the star if already in favorites', async () => {
    const { component, store } = mountComponent();
    let state = store.getState();
    expect(state.favorites.albums).toEqual([]);

    await waitFor(() => component.getByTestId(/add-remove-favorite/i).click());
    state = store.getState();
    expect(state.favorites.albums).toEqual([
      expect.objectContaining({
        artist: 'test artist',
        title: 'test album'
      })
    ]);

    await waitFor(() => component.getByTestId(/add-remove-favorite/i).click());
    state = store.getState();
    expect(state.favorites.albums).toEqual([]);
  });

  it('should play all tracks in the album after clicking play all', async () => {
    const { component, store } = mountComponent();
    await waitFor(() => component.getByText(/Play/i).click());

    const state = store.getState();
    expect(state?.queue?.queueItems).toEqual([
      expect.objectContaining({
        artist: 'test artist',
        name: 'test track 1'
      }),
      expect.objectContaining({
        artist: 'test artist',
        name: 'test track 2'
      }),
      expect.objectContaining({
        artist: 'test artist',
        name: 'test track 3'
      })
    ]);
    expect(state.player.playbackStatus).toEqual('PLAYING');
  });

  it('should add album to queue after clicking the button', async () => {
    const { component, store } = mountComponent();
    await waitFor(() => component.getByTestId('more-button').click());
    await waitFor(() => component.getByText(/Add album to queue/i).click());
    const state = store.getState();

    expect(state?.queue?.queueItems).toEqual([
      expect.objectContaining({
        artist: 'test artist',
        name: 'test track 1'
      }),
      expect.objectContaining({
        artist: 'test artist',
        name: 'test track 2'
      }),
      expect.objectContaining({
        artist: 'test artist',
        name: 'test track 3'
      })
    ]);
  });

  it('should add album to downloads after clicking the button', async () => {
    const { component, store } = mountComponent();
    await waitFor(() => component.getByTestId('more-button').click());
    await waitFor(() => component.getByText(/Download/i).click());
    const state = store.getState();
    expect(state.downloads).toEqual([
      {
        completion: 0,
        status: 'Waiting',
        track: expect.objectContaining({
          artist: 'test artist',
          name: 'test track 1',
          uuid: 'track-1-id'
        })
      },
      {
        completion: 0,
        status: 'Waiting',
        track: expect.objectContaining({
          uuid: 'track-2-id',
          artist: 'test artist',
          name: 'test track 2'
        })
      },
      {
        completion: 0,
        status: 'Waiting',
        track: expect.objectContaining({
          uuid: 'track-3-id',
          artist: 'test artist',
          name: 'test track 3'
        })
      }
    ]);
  });

  const mountComponent = mountedComponentFactory(
    ['/album/test-album-id'],
    buildStoreState()
      .withAlbumDetails()
      .withArtistDetails()
      .withPlugins()
      .withConnectivity()
      .build()
  );
});
