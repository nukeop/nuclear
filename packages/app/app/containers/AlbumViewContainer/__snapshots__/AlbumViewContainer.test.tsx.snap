// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Album view container should display an album 1`] = `
<DocumentFragment>
  <div
    class="main_layout_container"
  >
    <div
      class="album_view_container"
    >
      <div
        class="dimmable"
      >
        <div
          class="ui dimmer"
        >
          <div
            class="content"
          >
            <div
              class="ui loader"
            />
          </div>
        </div>
        <div
          class="album"
        >
          <div
            class="album_info_box"
          >
            <span
              class="nuclear small loader"
            >
              <span />
              <span />
              <span />
              <span />
            </span>
            <div
              class="album_details"
            >
              <div
                class="album_title"
              >
                <div
                  class="album_text"
                >
                  test album
                </div>
                <a
                  class="album_favorites_button_wrap"
                  data-testid="add-remove-favorite"
                  href="#"
                >
                  <i
                    aria-hidden="true"
                    class="star outline big icon"
                  />
                </a>
              </div>
              <div
                class="album_artist"
              >
                by 
                <a
                  href="#"
                >
                  test artist
                </a>
              </div>
              <div
                class="album_genre"
              >
                <label>
                  Genre:
                </label>
                genre 1, genre 2
              </div>
              <div
                class="album_year"
              >
                <label>
                  Year:
                </label>
                2001
              </div>
              <div
                class="album_tracks"
              >
                <label>
                  Tracks:
                </label>
                3
              </div>
              <div
                class="album_buttons"
              >
                <a
                  class="play_button"
                  href="#"
                >
                  <i
                    aria-hidden="true"
                    class="play icon"
                  />
                   Play
                </a>
                <a
                  class="more_button"
                  data-testid="more-button"
                  href="#"
                >
                  <i
                    aria-hidden="true"
                    class="ellipsis horizontal icon"
                  />
                </a>
              </div>
            </div>
          </div>
          <table
            class="album_tracklist"
          >
            <thead>
              <tr>
                <th
                  class="center position"
                >
                  <i
                    aria-hidden="true"
                    class="hashtag icon"
                  />
                </th>
                <th
                  class="left"
                >
                  Tracks:
                </th>
                <th
                  class="center duration"
                >
                  <i
                    aria-hidden="true"
                    class="clock outline icon"
                  />
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="track"
              >
                <td
                  class="track_number"
                />
                <td
                  class="track_name"
                >
                  test track 1
                </td>
                <td
                  class="track_duration"
                >
                  02:00
                </td>
              </tr>
              <tr
                class="track"
              >
                <td
                  class="track_number"
                />
                <td
                  class="track_name"
                >
                  test track 2
                </td>
                <td
                  class="track_duration"
                >
                  01:03
                </td>
              </tr>
              <tr
                class="track"
              >
                <td
                  class="track_number"
                />
                <td
                  class="track_name"
                >
                  test track 3
                </td>
                <td
                  class="track_duration"
                >
                  00:07
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Album view container should show that an album is still loading 1`] = `
<DocumentFragment>
  <div
    class="main_layout_container"
  >
    <div
      class="album_view_container"
    >
      <div
        class="dimmable"
      >
        <div
          class="ui active transition visible dimmer"
          style="display: flex;"
        >
          <div
            class="content"
          >
            <div
              class="ui loader"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
